core/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -505,16 +492,17 @@ public void deleteClosedIndex(String reason, IndexMetaData metaData, ClusterStat
     public void deleteIndexStore(String reason, IndexMetaData metaData, ClusterState clusterState, boolean closed) throws IOException {
         if (nodeEnv.hasNodeFile()) {
             synchronized (this) {
-                String indexName = metaData.getIndex().getName();
-                if (indices.containsKey(indexName)) {
-                    String localUUid = indices.get(indexName).indexUUID();
-                    throw new IllegalStateException("Can't delete index store for [" + indexName + "] - it's still part of the indices service [" + localUUid + "] [" + metaData.getIndexUUID() + "]");
+                Index index = metaData.getIndex();
+                if (hasIndex(index)) {
```

Comment:
on master we don't have shard locks. I was thinking that for example pending deletes can come back and delete an folder already created for a new index. This is already an issue so we can leave it as is and it will be fixed by making index level folder unique